iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java

```diff
@@ -244,6 +255,101 @@ public TSStatus handleMetaChanges(PipeHandleMetaChangePlan plan) {
     return new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
   }
 
+  public boolean hasExceptions(String pipeName) {
+    if (!pipeMetaKeeper.containsPipeMeta(pipeName)) {
+      return false;
+    }
+    PipeRuntimeMeta runtimeMeta = pipeMetaKeeper.getPipeMeta(pipeName).getRuntimeMeta();
+    Map<Integer, PipeRuntimeException> exceptionMap =
+        runtimeMeta.getDataNodeId2PipeRuntimeExceptionMap();
+    if (!exceptionMap.isEmpty()) {
+      return true;
+    }
+    AtomicBoolean hasException = new AtomicBoolean(false);
+    runtimeMeta
+        .getConsensusGroupId2TaskMetaMap()
+        .values()
+        .forEach(
+            pipeTaskMeta -> {
+              if (pipeTaskMeta.getExceptionMessages().iterator().hasNext()) {
+                hasException.set(true);
+              }
+            });
+    return hasException.get();
+  }
+
+  /**
+   * Clear the exceptions of a pipe locally after it starts successfully. If there are exceptions
+   * cleared, the messages will then be updated to all the nodes through
+   * PipeHandleMetaChangeProcedure.
+   *
+   * @param pipeName The name of the pipes to be clear exception
+   */
+  public void clearExceptions(String pipeName) {
+    if (!pipeMetaKeeper.containsPipeMeta(pipeName)) {
+      return;
+    }
+    PipeRuntimeMeta runtimeMeta = pipeMetaKeeper.getPipeMeta(pipeName).getRuntimeMeta();
+    runtimeMeta.setExceptionsClearTime(System.currentTimeMillis());
+    Map<Integer, PipeRuntimeException> exceptionMap =
+        runtimeMeta.getDataNodeId2PipeRuntimeExceptionMap();
+    if (!exceptionMap.isEmpty()) {
+      exceptionMap.clear();
+    }
+    runtimeMeta
+        .getConsensusGroupId2TaskMetaMap()
+        .values()
+        .forEach(
+            pipeTaskMeta -> {
+              if (pipeTaskMeta.getExceptionMessages().iterator().hasNext()) {
+                pipeTaskMeta.clearExceptionMessages();
+              }
+            });
+  }
+
+  /**
+   * Record the exceptions of all pipes locally if they encountered failure when pushing pipe meta.
+   * If there are exceptions recorded, the related pipes will be stopped, and the exception messages
+   * will then be updated to all the nodes through PipeHandleMetaChangeProcedure.
+   *
+   * @param respMap The responseMap after pushing pipe meta
+   * @return true if there are exceptions encountered
+   */
+  public boolean recordPushPipeMetaExceptions(Map<Integer, TPushPipeMetaResp> respMap) {
+    boolean hasException = false;
+    for (Map.Entry<Integer, TPushPipeMetaResp> respEntry : respMap.entrySet()) {
+      int dataNodeId = respEntry.getKey();
+      TPushPipeMetaResp resp = respEntry.getValue();
+      if (resp.getStatus().getCode() == TSStatusCode.PIPE_PUSH_META_ERROR.getStatusCode()) {
+        hasException = true;
+        if (!resp.isSetExceptionMessages()) {
```

Comment:
What if hasException is true and resp.isSetExceptionMessages() is false? What message will be sent back to user?

Reply:
Is this case handled by parsePushPipeMetaExceptionForPipe?